pulsar-client/src/main/java/org/apache/pulsar/client/impl/transaction/TransactionImpl.java

```diff
@@ -97,8 +104,8 @@ public enum State {
         this.registerSubscriptionMap = new ConcurrentHashMap<>();
         this.tcClient = client.getTcClient();
 
-        this.sendFutureList = new ArrayList<>();
-        this.ackFutureList = new ArrayList<>();
+        this.sendFuture = new CompletableFuture<>();
+        this.ackFuture = new CompletableFuture<>();
```

Comment:
It should be completed future or null?
Otherwise, if you don't have ack operation in this transaction, the allOpComplete method will never be completed.